Add Episode 2 exercises (#26)
czi-catalystproject/hub-champion-training: episodes/administer_hub.md

::::

::::{dropdown} Optimise server resources for your work
:open:

> In each scenario, select the most suitable option that makes best use out of the following server resources: *Shared Small: 1–4 CPU, 8–32 GB, Small: 4 CPU, 32 GB, Medium: 16 CPU, 128 GB, Large: 64 CPU, 512 GB*.

Author: ```suggestion
In each scenario, select the option that makes best use out of the following server resources: *Shared Small: 1–4 CPU, 8–32 GB RAM, Small: 4 CPU, 32 GB RAM, Medium: 16 CPU, 128 GB RAM, Large: 64 CPU, 512 GB RAM*.
```

> In each scenario, select the most suitable option that makes best use out of the following server resources: *Shared Small: 1–4 CPU, 8–32 GB, Small: 4 CPU, 32 GB, Medium: 16 CPU, 128 GB, Large: 64 CPU, 512 GB*.

Author: Resolved in https://github.com/czi-catalystproject/hub-champion-training/pull/26/commits/065d2f429a06442fd46d47322f72c26974739ea2

a. Run a software pipeline that has a single-thread task requiring 2GB of RAM and 60 tasks in total.

b. Create documentation and edit text files.

> c. Compile a software program that ran out of memory and failed after initially building on a *Shared Small* machine.

Reviewer: This one feels a little unclear, perhaps because there is no "correct" answer that exactly fits the framing of the exercise. Perhaps you could add it as a "stretch goal" question after the first three. E.g.

> Ruth wants to run a new program but does not know its maximum memory requirement. She tried compiling it a *Shared Small* machine, where it initially built but then ran out of memory and failed on execution. Which server configuration should Ruth choose to try to run the program again?

Author: I like this idea! Implemented in https://github.com/czi-catalystproject/hub-champion-training/pull/26/commits/7e9eacea83202d44f80bf0e1b25d4d3cdd48cb92

d. Preprocess a dataset consisting of 10,000 images, where each image is 16 MB in size.

:::{dropdown} Solution

> a. *Medium: 16 CPU, 128 GB* – the memory required is $ 60 \times 2 \, \mathrm{GB} = 120 \, \mathrm{GB}$, therefore the RAM on a *Medium* machine is sufficient.
> 
> b. *Shared Small: 1–4 CPU* – this is not a computationally intensive task so a machine shared with other users is sufficient.
> 
> c. *Small: 4 CPU, 32 GB* – the task failed on a *Shared Small* machine and we recommend sizing up incrementally if you are unsure of the server resources you need.
> 
> d. *Large: 64 CPU, 512 GB* – the total dataset size is $10000 \times 16 \, \mathrm{MB} = 160 000 \, \mathrm{MB} = 160 \, \mathrm{GB}$, which is too large for the *Medium* machine to hold in memory at once. Alternatively, you could chunk up your dataset and preprocess each batch on a smaller machine, which would be slower. Image preprocessing is often a CPU bound process in deep learning workflows, so a *Large* machine with more CPU cores is best for performance.

Author: ```suggestion
a. *Medium: 16 CPU, 128 GB RAM* – the memory required is $ 60 \times 2 \, \mathrm{GB} = 120 \, \mathrm{GB}$, therefore the RAM on a *Medium* machine is sufficient.

b. *Shared Small: 1–4 CPU, 8–32 GB RAM* – this is not a computationally intensive task so a machine shared with other users is sufficient.

c. *Small: 4 CPU, 32 GB RAM* – the task failed on a *Shared Small* machine and we recommend sizing up incrementally if you are unsure of the server resources you need.

d. *Large: 64 CPU, 512 GB RAM* – the total dataset size is $10000 \times 16 \, \mathrm{MB} = 160 000 \, \mathrm{MB} = 160 \, \mathrm{GB}$, which is too large for the *Medium* machine to hold in memory at once. Alternatively, you could chunk up your dataset and preprocess each batch on a smaller machine, which would be slower. Image preprocessing is often a CPU bound process in deep learning workflows, so a *Large* machine with more CPU cores is best for performance.
```

:::

::::

::::{dropdown} User server management
:open:

> Which of the following statements are *True* or *False*?

Reviewer: ```suggestion
Are the following statements *True* or *False*?
```
